File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {

    ListNode prev;
    ListNode tempNext;
    ListNode curr;
    public ListNode reverseList(ListNode head) {
        curr = head;
        while (curr != null){
            //(0,1,2,3)
            // curr =0,1,2,3
        ListNode nextTemp = curr.next; //t=1,2,3,nl
        curr.next = prev; //cn=nul,0,1,2
        prev = curr; //p=0,1,2,3
        curr = nextTemp; //c=1,2,3,null


            // curr.next = tempNext;
            // curr.next = prev;
        }
        return prev;
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # dummy and left are 0
        # right = n -> while n>0 and right(notnull), r = r.next, n-=1
        # loop though, while right: l.next, r.next
        # jump left to 1 more, l.nex = l.nex.nex
        # everything is attached to dummy viz attached to head, so dummy.nex

        dum = ListNode(0,head)
        left = dum
        right = head

        while n >0 and right:
            right = right.next
            n-=1
        
        while right: 
        #and right.next: not including because gave (1,3,4).here,r is 3 before loop and after running once, r.nex becomes null and then it won't run, leaving l as 1, then l.nex = l.nex.nex skips 2 instead of 3. 
            left = left.next
            right = right.next
        left.next = left.next.next

        return dum.next



        
       
        
            
                
        
        